feat(diff): merge adjacent blocks using inline:word

HarshK97 opened the pull request:
## Problem:
When using 'diffopt=inline:word', lines were excessively fragmented with punctuation creating separate highlight blocks, making it harder to read the diffs.
<img width="3132" height="2038" alt="Screenshot 2025-12-27 at 10 29 52 PM" src="https://github.com/user-attachments/assets/ed7cfa59-d5f1-4d76-9394-a30b89a216e1" />

## Solution:
This PR replaces the old behavior of fragmentation in between words separated by 'spaces' or 'punctuation' to joining those words together.  
Which when enabled `inline:word` provides continuous block of diff.
- This creates more readable diffs and closely matches GitHub's own diff display.

<img width="2940" height="1912" alt="36025" src="https://github.com/user-attachments/assets/a2d64e67-0278-4fba-91cb-87275699ddbe" />

## Usage
Enable with:
```
:set diffopt+=inline:word
```

## Implementation
- Adds the **diff_refine_inline_word_highlight()** implementation in C to merge word fragments separated by punctuation or whitespace (≤5 bytes)
- Option parsing, docs, and completion support included.

## Tests
- Added/updated functional tests to cover:
  - Updated tests for `inline:word` to support the new behavior by default
  - Edge cases: multi-byte, iwhite, iwhiteeol, punctuation, 3-buffer diffs
  
## Documentation
- Updated `:help diff`, options, and news/changelog/documentation.

This creates more readable and is similar to github diff behaviour with backward compactibility.

Fixes #37019

HarshK97 committed:
feat(diff): added mergeblocks option to reduce inline diff fragmentation

Added mergeblocks option to separate the old functionality and the new
merging functionality, which helps in backward-compactibility.

The option is off by default and can be enabled with:
  :set diffopt+=inline:word,mergeblocks

Implementation:
- Added DIFF_MERGE_BLOCKS flag (0x40000) to diff.c
- Added diff_refine_inline_word_highlight() function to merge adjacent
  blocks on the same line
- Parse 'mergeblocks' from diffopt string
- Apply merging only when inline:word is active
- Uses multi-pass approach (up to 4 iterations) to handle cascading
  merges

Testing:
- Added comprehensive test coverage for both old and new behavior
- Tests verify backward compatibility with existing inline:word behavior
- Covers multi-byte characters (emojis, CJK, Greek), iwhite, iwhiteeol
- Tests 3-buffer diff scenarios and toggle behavior (diffopt/diffthis)

Documentation:
- Updated runtime/doc/options.txt with mergeblocks flag description
- Added usage notes in runtime/doc/diff.txt
- Included news.txt entry for user-facing changelog
- Updated test_options.vim for option completion

Fixes #37019

justinmk commented on runtime/doc/news.txt:
```diff
@@ -34,6 +34,9 @@ LSP
 OPTIONS
 
 • `'completefuzzycollect'` and `'isexpand'` have been removed.
+• `'mergeblocks'` option for 'diffopt':  Merges adjacent diff blocks separated
```
This is not a breaking change. Should be in the news-features section of news.txt.

HarshK97 replied:
Ok, but with now I am focusing on making this the old behavior, should I remove about mergeblocks and add about new behavior in breaking change?

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2295,6 +2295,13 @@ A jump table for the options with a short description can be found at |Q_op|.
 		vertical	Start diff mode with vertical splits (unless
 				explicitly specified otherwise).
 
+		mergeblocks	When using 'inline:word', merge adjacent diff
```
why is this at the end? the flags are in alphanum order.

HarshK97 replied:
Oh I was unaware of it being alphanum order, that's why I did this

HarshK97 committed:
feat(diff): merge adjacent blocks for inline:word by default

When using 'diffopt' with 'inline:word', adjacent diff blocks separated
by small gaps (≤3 characters of whitespace or punctuation) are now
automatically merged when the combined change size is significant
(>= gap * 4). This creates more readable diffs similar to GitHub's
display and reduces visual fragmentation.

This behavior is now the default for inline:word mode, matching how
inline:char already merges blocks to improve readability.

Implementation:
- Added diff_refine_inline_word_highlight() function to merge adjacent
  blocks on the same line
- Uses multi-pass approach (up to 4 iterations) to handle cascading merges
- Called automatically when DIFF_INLINE_WORD flag is active

Testing:
- Updated test expectations in test/functional/ui/diff_spec.lua
- Tests verify merging with multi-byte characters (emojis, CJK, Greek)
- Covers iwhite, iwhiteeol options and 3-buffer diff scenarios
- All tests show improved merged highlighting behavior

Documentation:
- Added feature entry in runtime/doc/news.txt

Fixes #37019

HarshK97 commented:
@lewis6991 I am very grateful that you like this change that I made to the diffopt, and I would love to make this configurable!

HarshK97 committed:
feat(diff): merge adjacent blocks with inline:word using configurable wordgaps

When using 'diffopt' with 'inline:word', adjacent diff blocks separated
by small gaps can now be merged in a configurable way using 'wordgaps'.

With this new behaviour for inline:word mode, users can explicitly
control how small gaps between word-level diff blocks are handled.

Implementation:
- Added 'wordgaps:' option to configure inline:word behaviour

Testing:
- Tests for 'wordgaps' in tests/functional_tests/diff_spec.lua are
pending

Documentation:
- Added option entry for 'wordgaps' in src/nvim/options.lua
- Added feature entry in runtime/doc/news.txt

HarshK97 committed:
fix(diff): fixed a bug in diff_refine_inline_word_highlight()

old implementation was checking number of tokens instead of bytes or
character which resulted in problems when ever the gap_size was bigger
than 3
Found at the time of testing

HarshK97 committed:
test(diff): added tests for wordgap option in diffopt for inline:word

Added Compresensive tests like
- wordgap:0 disables merging
- wordgap:2 merges gaps up to 2 bytes
- wordgap:5 merges larger gaps but not that big
- wordgap shows difference between gap thresholds
- wordgap works with underscores
- wordgap works with punctuation
- wordgap maximum value is 5
- wordgap respects iwhite option
- wordgap works with multi-byte characters
- wordgap can be toggled dynamically

HarshK97 committed:
feat(diff): removed 'wordgap' options and 5 as default for diff_word_gap

- Removed the 'wordgap' option from diffopt
- Removed the docs related about 'wordgap'
- Removed tests related about 'wordgap'
- For more info look at the vim pr for the same feature: https://github.com/vim/vim/pull/19098

HarshK97 committed:
docs(options): updated options.lua for inline:word

justinmk commented:
> due to ychin's comment on [vim/vim#19098 (comment)](https://github.com/vim/vim/pull/19098#issuecomment-3739713871),

Oh I didn't know this was being upstreamed. Ok, that really simplifies things. Will be better if Vim merges it first then please update here. I agree with ychin's comments in the Vim issue.

zeertzjq committed:
vim-patch:9.2.0174: diff: inline word-diffs can be fragmented

Problem:  When using 'diffopt=inline:word', lines were excessively
          fragmented with punctuation creating separate highlight
          blocks, making it harder to read the diffs.
Solution: Added 'diff_refine_inline_word_highlight()' to merge
          adjacent diff blocks that are separated by small gaps of
          non-word characters (up to 5 bytes by default) (HarshK97).

When using inline:word diff mode, adjacent changed words separated by
punctuation or whitespace are now merged into a single highlight block
if the gap between them contains fewer than 5 non-word characters.

This creates more readable diffs and closely matches GitHub's own diff
display.

closes: vim/vim#19098

https://github.com/vim/vim/commit/42c6686c78d39843f71dba989a8ea59bc6975132

Co-authored-by: HarshK97 <harshkapse1234@gmail.com>
